### ISSUE TITLE
Enabling constraint match statistics during benchmarking leads to missleading warning

### ISSUE THREAD

**Issue by jason076** [TimefoldAI/timefold-solver#361]
**Describe the bug**
Disclaimer: I dont't know if this a bug or just missing documentation.

Activating `CONSTRAINT_MATCH_TOTAL_BEST_SCORE` or `CONSTRAINT_MATCH_TOTAL_STEP_SCORE` leads to the following waring message: 

```
The metrics [timefold.solver.constraint.match.step.score, timefold.solver.constraint.match.best.score] cannot function properly because ConstraintMatches are not supported on the ScoreDirector.
```

I am using the `BavetConstraintStreamScoreDirector` after debugging I found out that this warning is printed because the environment_mode was non asserted. In the docs there is nothing mentioned about that the environment mode has to be asserted https://timefold.ai/docs/timefold-solver/latest/benchmarking-and-tweaking/benchmarking-and-tweaking#benchmarkReportConstraintMatchTotalBestScoreOverTimeStatistic. Furthermore should't constraint matches be available in non asserted environment modes? If I execute the benchmark and ignore the warning I get a valid benchmark report. So maybe the check if the metric is available is wrong here?

**Expected behavior**
Get no error message then activating CONSTRAINT_MATCH_TOTAL_BEST_SCORE with a constraint stream ScoreDirector.

**Actual behavior**
The above mentioned warning is printed if the environment mode is set to non asserted. 

**To Reproduce**

1. Create a PlanningModel using ConstraintStreams
2. Enable CONSTRAINT_MATCH_TOTAL_BEST_SCORE for a benchmark
3. Set environment mode to REPRODUCIBLE or NON_REPRODUCIBLE for the benchmark solver
4. Execute Benchmark

## Environment

**Timefold Solver Version or Git ref**: 0.8.41

**Output of `java -version`:**
```
openjdk 17.0.8.1 2023-08-22 LTS
OpenJDK Runtime Environment Corretto-17.0.8.8.1 (build 17.0.8.1+8-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.8.8.1 (build 17.0.8.1+8-LTS, mixed mode, sharing)
```

**Output of `uname -a` or `ver`:**
```
23.0.0 Darwin Kernel Version 23.0.0: Fri Sep 15 14:43:05 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T6020 arm64
```

**Comment by triceo** [comment]
Thanks for reporting, @jason076.

Currently the public API only provides two ways to enable constraint matching:

- In asserted modes
- or in `SolutionManager.explain(...)`.

Unless you've done one of the two things, constraint matches will not be enabled. This is intentional - constraint matching has considerable performance impact, and therefore it is opt-in.

I have to admit I'm not sure how this interacts with the benchmarker. I totally agree that, when these statistics are enabled, constraint matching needs to be enabled automatically. If that doesn't happen already, this is clearly a bug. And if it does happen, then the message shouldn't be shown.

Will investigate when time permits.

**Comment by triceo** [comment]
In the end, the problem was a bit of both.

The benchmarker was already working correctly, and incorrectly printing out the warning.
The solver (when benchmarker is not used) was not working correctly with those metrics, and the warning was warranted.

Both have been fixed in 1.4.0, which is likely going to be released next month.
